File: exercises/01-nestjs-refactor/poor-code.ts
import { Controller, Get, Query } from '@nestjs/common';

@Controller('users')
export class UsersController {
  private users = [
    { id: 1, name: 'Andres Rivero', role: 'Frontend Developer' },
    { id: 2, name: 'Carlos Perez', role: 'Backend Developer' },
    { id: 3, name: 'Gabriel Silva', role: 'Full Stack Developer' },
    { id: 4, name: 'Maria Gomez', role: 'Mobile Developer' },
  ];

  @Get('search')
  searchUsers(@Query('q') query: string) {
    const results = this.users.filter((user) =>
      user.name.toLowerCase().includes(query.toLowerCase()),
    );

    return {
      count: results.length,
      data: results,
    };
  }
}

File: exercises/02-flutter-refactor/poor_code.dart
import 'dart:async';
import 'package:flutter/material.dart';

class SearchPage extends StatefulWidget {
  @override
  State<SearchPage> createState() => _SearchPageState();
}

class _SearchPageState extends State<SearchPage> {
  TextEditingController controller = TextEditingController();
  Timer? debounce;

  List<String> results = [];

  final users = [
    'Andres Rivero',
    'Carlos Perez',
    'Gabriel Silva',
    'Maria Gomez',
  ];

  void onSearchChanged(String query) {
    debounce = Timer(Duration(milliseconds: 500), () {
      setState(() {
        results = users
            .where((user) => user.toLowerCase().contains(query.toLowerCase()))
            .toList();
      });
    });
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: TextField(
        controller: controller,
        onChanged: onSearchChanged,
      ),
    );
  }
}
